botocore/stub.py

```diff
@@ -0,0 +1,164 @@
+        :type service_response: dict
+        """
+        self._add_response(method, service_response)
+
+    def _add_response(self, method, service_response, http_response=None):
+        if not hasattr(self.client, method):
+            raise ValueError(
+                "Client %s does not have method: %s"
+                % (self.client.meta.service_model.service_name, method))
+
+        if http_response is None:
+            http_response = Response()
+            http_response.status_code = 200
+            http_response.reason = 'OK'
+
+        operation_name = self.client.meta.method_to_api_mapping.get(method)
+        self._validate_response(operation_name, service_response)
+
+        response = (operation_name, (http_response, service_response))
+        self._queue.append(response)
+
+    def add_client_error(self, method, service_error_code='',
+                         service_message='', http_status_code=400,
+                         http_reason=''):
+        """
+        Adds a ClientError to the response queue.
+
+        :param method: The name of the service method to return the error on.
+        :type method: str
+
+        :param service_error_code: The service error code to return,
+                                   e.g. NoSuchBucket
+        :type service_error_code: str
+
+        :param service_message: The service message to return, e.g.
+                        'The specified bucket does not exist.'
+        :type service_message: str
+
+        :param http_status_code: The HTTP status code to return, e.g. 404, etc
+        :type http_status_code: int
+
+        :param http_reason: The HTTP reason to return, e.g. 'Not Found'
+        :type http_reason: str
+        """
+        http_response = Response()
+        http_response.status_code = http_status_code
+        http_response.reason = http_reason
+
+        # We don't look to the model to build this because the caller would need
+        # to know the details of what the HTTP body would need to look like.
+        parsed_response = {
+            'ResponseMetadata': {'HTTPStatusCode': http_status_code},
+            'Error': {
+                'Message': service_message,
+                'Code': service_error_code
+            }
+        }
+
+        operation_name = self.client.meta.method_to_api_mapping.get(method)
+        response = (operation_name, (http_response, parsed_response))
+        self._queue.append(response)
+
+    def assert_no_pending_responses(self):
+        """
+        Asserts that all expected calls were made.
+        """
+        remaining = len(self._queue)
+        if remaining != 0:
+            raise AssertionError("%d responses remaining in queue." % remaining)
+
+    def _get_response_handler(self, model, **kwargs):
+        if len(self._queue) < 1:
+            raise StubResponseError(
+                operation_name=model.name,
+                reason='No responses remaining.')
```

Comment:
What about updating this error message to also include something like `Unexpected API call: <the API call + args>` to give more context to the user about what went wrong.  A  misconfigured mock can be tedious to track down if the error messages don't give a lot of detail.